src/v/rpc/reconnect_transport.cc

```diff
@@ -72,33 +72,40 @@ reconnect_transport::reconnect(clock_type::time_point connection_timeout) {
-            1,
-            connection_timeout_duration,
-            [this, connection_timeout] {
-                if (is_valid()) {
-                    return ss::make_ready_future<ret_t>(&_transport);
-                }
-                vlog(
-                  rpclog.trace,
-                  "connecting to {}",
-                  _transport.server_address());
-                return _transport.connect(connection_timeout)
-                  .then_wrapped([this](ss::future<> f) {
-                      try {
-                          f.get();
-                          rpclog.debug(
-                            "connected to {}", _transport.server_address());
-                          _backoff_policy.reset();
-                          return ss::make_ready_future<ret_t>(&_transport);
-                      } catch (...) {
-                          _backoff_policy.next_backoff();
-                          rpclog.trace(
-                            "error reconnecting {}", std::current_exception());
-                          return ss::make_ready_future<ret_t>(
-                            errc::disconnected_endpoint);
-                      }
-                  });
+                   _connected_sem,
+                   1,
+                   connection_timeout_duration,
+                   [this, connection_timeout] {
+                       if (is_valid()) {
+                           return ss::make_ready_future<ret_t>(&_transport);
+                       }
+                       vlog(
+                         rpclog.trace,
+                         "connecting to {}",
+                         _transport.server_address());
+                       return _transport.connect(connection_timeout)
+                         .then_wrapped([this](ss::future<> f) {
+                             try {
+                                 f.get();
+                                 rpclog.debug(
+                                   "connected to {}",
+                                   _transport.server_address());
+                                 _backoff_policy.reset();
+                                 return ss::make_ready_future<ret_t>(
+                                   &_transport);
+                             } catch (...) {
+                                 _backoff_policy.next_backoff();
+                                 rpclog.trace(
+                                   "error reconnecting {}",
+                                   std::current_exception());
+                                 return ss::make_ready_future<ret_t>(
+                                   errc::disconnected_endpoint);
+                             }
+                         });
+                   })
+            .handle_exception_type([](const ss::named_semaphore_timed_out&) {
+                rpclog.trace("timeout waiting for RPC reconnect semaphore");
+                return ss::make_ready_future<ret_t>(errc::connection_timeout);
```

Comment:
Is the only timeout exception we expect above from the `with_semaphore` call? Can `transport::connect()` throw, or does that get swallowed in the above catch -- if the latter, should we be handling it as a `connection_timeout`?